### ISSUE TITLE
Stabilize gradient for unfolded Tensor

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#20806]
## 🚀 Feature

I propose a `stabilize_grad` flag for the `Tensor.unfold` method.

## Motivation

Imagine you want to calculate a loss between different patches of two images. Consider the following snippet:
```python
height = 8
width = 8
size = 3
step = 2

input_unstabilized = torch.ones(1, 1, height, width).requires_grad_(True)
target = torch.zeros(1, 1, height, width).detach()

input_unstabilized_patches = input_unstabilized
target_patches = target
for dim in range(2, 4):
    input_unstabilized_patches = input_unstabilized_patches.unfold(dim, size, step)

    target_patches = target_patches.unfold(dim, size, step)

loss = torch.sum((input_unstabilized_patches - target_patches) ** 2.0)
loss.backward()
print(input_unstabilized.grad[0, 0, :, :])
```
This results in:
```
tensor([[2., 2., 4., 2., 4., 2., 2., 0.],
        [2., 2., 4., 2., 4., 2., 2., 0.],
        [4., 4., 8., 4., 8., 4., 4., 0.],
        [2., 2., 4., 2., 4., 2., 2., 0.],
        [4., 4., 8., 4., 8., 4., 4., 0.],
        [2., 2., 4., 2., 4., 2., 2., 0.],
        [2., 2., 4., 2., 4., 2., 2., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.]])
```
As you can see, some elements receive a higher gradient update, simply because they appear in multiple patches. This effect might lead to unwanted results.

## Pitch

With `stabilize_grad=True` the `unfold` method should track how often each element appears in a patch and in the backward pass average the gradient with this number.

## Alternatives

Right now I achieve this functionality as follows:
```python
class _UnfoldGradStabilizer(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, dim, size, step):
        ctx.needs_stabilizing = step < size
        if ctx.needs_stabilizing:
            stabilizer = torch.zeros_like(input)
            item = [slice(None) for _ in range(input.dim())]
            for idx in range(0, stabilizer.size()[dim] - size, step):
                item[dim] = slice(idx, idx + size)
                stabilizer[item].add_(1.0)

            # clamping to avoid zero division
            ctx.save_for_backward(torch.clamp(stabilizer, min=1.0))
        return input

    @staticmethod
    def backward(ctx, grad_output):
        if ctx.needs_stabilizing:
            stabilizer, = ctx.saved_tensors
            grad_input = grad_output / stabilizer
        else:
            grad_input = grad_output.clone()
        return grad_input, None, None, None


unfold_grad_stabilizer = _UnfoldGradStabilizer.apply


def stable_unfold(input, dim, size, step):
    return unfold_grad_stabilizer(input, dim, size, step).unfold(dim, size, step)


input_stabilized = torch.ones(1, 1, height, width).requires_grad_(True)

input_stabilized_patches = input_stabilized
for dim in range(2, 4):
    input_stabilized_patches = stable_unfold(input_stabilized_patches, dim, size, step)

loss = torch.sum((input_stabilized_patches - target_patches) ** 2.0)
loss.backward()
print(input_stabilized.grad[0, 0, :, :])
```
This results in
```
tensor([[2., 2., 2., 2., 2., 2., 2., 0.],
        [2., 2., 2., 2., 2., 2., 2., 0.],
        [2., 2., 2., 2., 2., 2., 2., 0.],
        [2., 2., 2., 2., 2., 2., 2., 0.],
        [2., 2., 2., 2., 2., 2., 2., 0.],
        [2., 2., 2., 2., 2., 2., 2., 0.],
        [2., 2., 2., 2., 2., 2., 2., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.]])
```

I have not yet timed itin a larger context, but I suspect the creation of the `stabilizer` with a `for` loop in every forward pass might be slow. If you can think of a better way, please let me know.

**Comment by ezyang** [comment]
This sounds plausible. But I wonder if there is a way to do this generally for any viewing operations, not just unfold.

**Comment by pmeier** [comment]
Which operations are you referring to? It is not needed for `Tensor.view()` since the number of elements stays constant, whereas `Tensor.unfold` might change it. From the above example:
```python
print(input.numel())
print(input_patches.numel())
```
This results in
```
64
81
```

**Comment by fmassa** [comment]
But this is not a valid gradient anymore, right?

**Comment by pmeier** [comment]
What do you mean by _valid_? It would be stabilized / normalized / averaged / (whatever term you prefer) compared to the current behaviour.

**Comment by fmassa** [comment]
By valid I mean it doesn't correspond to the notion of a gradient per see, right? Overlapping elements should contribute to the gradients by summing them.

What I'm trying to say is, if you do finite differences, the gradient check wont match I think, except if you change the behavior during forward to divide the pixels of the output image by the number of overlapping regions

**Comment by pmeier** [comment]
You are right. I didn't properly think this through. Thanks for the heads up.
